### PR TITLE
Added support for InfraVnet connectivity in multitenancy CNI

### PR DESCRIPTION
**Pull request by tamilmani1989**
**What this PR does / why we need it**:
Added support for InfraVnet connectivity in multitenancy CNI. Added separate package for snat and infravnet

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: cni/network/network.go [hunk 1]
```diff
@@ -553,10 +580,20 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
 		return err
 	}
 
-	// Call into IPAM plugin to release the endpoint's addresses.
-	nwCfg.Ipam.Subnet = nwInfo.Subnets[0].Prefix.String()
-	for _, address := range epInfo.IPAddresses {
-		nwCfg.Ipam.Address = address.IP.String()
+	if !nwCfg.MultiTenancy {
```
Comment (sharmasushant) [inline cni/network/network.go]: bugfix

File: cni/network/network.go [hunk 2]
```diff
@@ -553,10 +580,20 @@ func (plugin *netPlugin) Delete(args *cniSkel.CmdArgs) error {
...
+		for _, address := range epInfo.IPAddresses {
+			nwCfg.Ipam.Address = address.IP.String()
+			err = plugin.DelegateDel(nwCfg.Ipam.Type, nwCfg)
+			if err != nil {
+				err = plugin.Errorf("Failed to release address: %v", err)
+				return err
+			}
+		}
+	} else if epInfo.EnableInfraVnet {
+		nwCfg.Ipam.Subnet = nwInfo.Subnets[0].Prefix.String()
```
Comment (sharmasushant) [inline cni/network/network.go]: Get the Subnet using information from CNS. That will be same in windows and linux.

File: cni/network/network_windows.go [hunk 3]
```diff
@@ -63,10 +63,13 @@ func handleConsecutiveAdd(containerId, endpointId string, nwInfo *network.Networ
...
 }
 
+func addInfraRoutes(azIpamResult *cniTypesCurr.Result, result *cniTypesCurr.Result, epInfo *network.EndpointInfo) {
+}
+
 func setNetworkOptions(cnsNwConfig *cns.GetNetworkContainerResponse, nwInfo *network.NetworkInfo) {
 }
 
-func setEndpointOptions(cnsNwConfig *cns.GetNetworkContainerResponse, epInfo *network.EndpointInfo, vethName string) {
+func setEndpointOptions(cnsNwConfig *cns.GetNetworkContainerResponse, infraIPNet *net.IPNet, epInfo *network.EndpointInfo, vethName string) {
```
Comment (sharmasushant) [inline cni/network/network_windows.go]: remove the new parameter from windows code

File: network/ovs_networkclient_linux.go [hunk 4]
```diff
@@ -183,186 +97,17 @@ func (client *OVSNetworkClient) DeleteBridge() error {
...
-	err := netlink.DeleteLink(snatBridgeName)
-	if err != nil {
-		log.Printf("[net] Failed to delete bridge %v, err:%v.", snatBridgeName, err)
-	}
-
-	return err
-}
-
 func (client *OVSNetworkClient) AddL2Rules(extIf *externalInterface) error {
 	//primary := extIf.IPAddresses[0].IP.String()
```
Comment (sharmasushant) [inline network/ovs_networkclient_linux.go]: remove dead code

File: network/ovsinfravnet/infravnet_windows.go [hunk 5]
```diff
@@ -0,0 +1 @@
+package ovsinfravnet
```
Comment (sharmasushant) [inline network/ovsinfravnet/infravnet_windows.go]: check if code compiles on windows

**Comment by tamilmani1989** [comment]
@sharmasushant addressed review comments.

**Pull request merged**